packages/@aws-cdk/aws-stepfunctions-tasks/lib/sagemaker-task-base-types.ts

```diff
@@ -218,6 +222,148 @@ export interface MetricDefinition {
     readonly regex: string;
 }
 
+/**
+ * Specifies a location in a S3 Bucket.
+ *
+ * @experimental
+ */
+export interface IS3Location {
+    /** The URI of the location in S3 */
+    readonly uri: string;
+
+    /** Grants read permissions to the S3 location. */
+    grantRead(grantee: iam.IGrantable): void;
+    /** Grants write permissions to the S3 location. */
+    grantWrite(grantee: iam.IGrantable): void;
+}
+
+/**
+ * Constructs `IS3Location` objects.
+ */
+export class S3Location {
+    /**
+     * An `IS3Location` built with a determined bucket and key prefix.
+     *
+     * @param bucket    is the bucket where the objects are to be stored.
+     * @param keyPrefix is the key prefix used by the location.
+     */
+    public static inBucket(bucket: s3.IBucket, keyPrefix: string): IS3Location {
+        return {
+            uri: bucket.urlForObject(keyPrefix),
+            grantRead: (grantee) => bucket.grantRead(grantee, keyPrefix + '*'),
+            grantWrite: (grantee) => bucket.grantWrite(grantee, keyPrefix + '*'),
+        };
+    }
+
+    /**
+     * An `IS3Location` determined fully by a JSON Path from the task input.
+     *
+     * Due to the dynamic nature of those locations, the IAM grants that will be set by `grantRead` and `grantWrite`
+     * apply to the `*` resource.
+     *
+     * @param expression the JSON expression resolving to an S3 location URI.
+     */
+    public static fromJsonExpression(expression: string): IS3Location {
+        return {
+            uri: sfn.Data.stringAt(expression),
+            grantRead: grantee => grantee.grantPrincipal.addToPolicy(new iam.PolicyStatement({
+                actions: ['s3:GetObject', `s3:ListBucket`],
+                resources: ['*']
+            })),
+            grantWrite: grantee => grantee.grantPrincipal.addToPolicy(new iam.PolicyStatement({
+                actions: ['s3:PutObject'],
+                resources: ['*']
+            })),
+        };
+    }
+
+    private constructor() { }
+}
+
+export interface IDockerImage {
+    readonly name: string;
+    grantRead(grantee: iam.IGrantable): void;
```

Comment:
Might make more sense to follow the `bind()` pattern that we try to use everywhere.